server/src/main/java/org/elasticsearch/common/geo/GeometryTreeWriter.java

```diff
@@ -53,29 +55,56 @@ public GeometryTreeWriter(Geometry geometry, CoordinateEncoder coordinateEncoder
         this.coordinateEncoder = coordinateEncoder;
         this.centroidCalculator = new CentroidCalculator();
         builder = new GeometryTreeBuilder(coordinateEncoder);
-        geometry.visit(builder);
+        if (geometry.type() == ShapeType.GEOMETRYCOLLECTION) {
+            for (Geometry shape : (GeometryCollection<?>) geometry) {
+                shape.visit(builder);
+            }
+        } else {
+            geometry.visit(builder);
+        }
     }
 
-    public Extent extent() {
+    @Override
+    public Extent getExtent() {
         return new Extent(builder.top, builder.bottom, builder.negLeft, builder.negRight, builder.posLeft, builder.posRight);
     }
 
+    @Override
+    public ShapeType getShapeType() {
+        return ShapeType.GEOMETRYCOLLECTION;
+    }
+
+    @Override
+    public CentroidCalculator getCentroidCalculator() {
+        return centroidCalculator;
+    }
+
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         // only write a geometry extent for the tree if the tree
         // contains multiple sub-shapes
-        boolean prependExtent = builder.shapeWriters.size() > 1;
+        boolean multiShape = builder.shapeWriters.size() > 1;
         Extent extent = null;
         out.writeInt(coordinateEncoder.encodeX(centroidCalculator.getX()));
         out.writeInt(coordinateEncoder.encodeY(centroidCalculator.getY()));
-        if (prependExtent) {
+        if (multiShape) {
             extent = new Extent(builder.top, builder.bottom, builder.negLeft, builder.negRight, builder.posLeft, builder.posRight);
         }
         out.writeOptionalWriteable(extent);
         out.writeVInt(builder.shapeWriters.size());
-        for (ShapeTreeWriter writer : builder.shapeWriters) {
-            out.writeEnum(writer.getShapeType());
-            writer.writeTo(out);
+        if (multiShape) {
+            for (ShapeTreeWriter writer : builder.shapeWriters) {
+                try(BytesStreamOutput bytesStream = new BytesStreamOutput()) {
```

Comment:
Would it be just as readable and slightly more efficient to re-use one `BytesStreamOutput` and reset it?

Reply:
As we discussed I would like to figure out how to optimize this code here and in other places. I have a couple of ideas.

Reply:
fine with punting this outside of this PR 👍